kernel/src/main/java/org/kframework/lsp/KTextDocumentService.java

```diff
@@ -0,0 +1,194 @@
+    public CompletableFuture<Either<List<CompletionItem>, CompletionList>> completion(CompletionParams position) {
+        return CompletableFuture.supplyAsync(() -> {
+            List<DefinitionItem> fileDi = new ArrayList<>(elems.getOrDefault(position.getTextDocument().getUri(), List.of()));
+            fileDi.addAll(elems.get(domains.toString()));
+            fileDi.addAll(elems.get(kast.toString()));
+            this.clientLogger.logMessage("Operation '" + "text/completion: " + position.getTextDocument() + " #di: " + fileDi.size());
+            return Either.forLeft(getCompletionItems(fileDi));
+        });
+    }
+
+    static Pattern ptrn = Pattern.compile("[a-zA-Z0-9#]+");
+
+    private static List<CompletionItem> getCompletionItems(List<DefinitionItem> dis) {
+        List<CompletionItem> lci = new ArrayList<>();
+        dis.stream().filter(i -> i instanceof Module)
+                .map(m -> ((Module) m))
+                .forEach(m -> m.getItems().stream()
+                        .filter(mi -> mi instanceof Syntax)
+                        .map(s -> ((Syntax) s))
+                        .forEach(s -> s.getPriorityBlocks()
+                                .forEach((pb -> pb.getProductions()
+                                        .forEach(p -> p.getItems().stream()
+                                                .filter(pi -> pi instanceof Terminal)
+                                                .map(t -> (Terminal) t)
+                                                .forEach(t -> {
+                                                    if (ptrn.matcher(t.getTerminal()).matches()) {
+                                                        CompletionItem completionItem = new CompletionItem();
+                                                        completionItem.setLabel(t.getTerminal());
+                                                        completionItem.setInsertText(t.getTerminal());
+                                                        completionItem.setDetail("module " + m.getName());
+                                                        String doc = "syntax ";
+                                                        doc += !s.getParams().isEmpty() ?
+                                                                "{" + s.getParams().stream().map(Sort::toString).collect(Collectors.joining(", ")) + "} " : "";
+                                                        doc += s.getDeclaredSort() + " ::= ";
+                                                        doc += p.toString();
+                                                        completionItem.setDocumentation(doc);
+                                                        completionItem.setKind(CompletionItemKind.Snippet);
+                                                        lci.add(completionItem);
+                                                    }
+                                                }))))));
+
+        return lci;
+    }
```

Comment:
Can we break this code down somewhat? It's a bit tricky to follow that many nested stream operations

Reply:
Extracting the core and annotating the tree structure definitely helps. If this is the best way to do this one-off traversal then I'm fine with it as is.